src/server/scripts/Zandalar/Underrot/boss_cragmaw_the_infested.cpp

```diff
@@ -0,0 +1,377 @@
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AreaTrigger.h"
+#include "AreaTriggerAI.h"
+#include "CreatureAI.h"
+#include "CreatureAIImpl.h"
+#include "InstanceScript.h"
+#include "ObjectAccessor.h"
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
+#include "SpellScript.h"
+#include "SpellAuras.h"
+#include "SharedDefines.h"
+#include "TemporarySummon.h"
+#include "underrot.h"
+
+enum CragmawSpells
+{
+    // Cragmaw
+    SPELL_CRAWG_EATING                  = 279156,
+    SPELL_TANTRUM_ENERGY_BAR            = 271771,
+    SPELL_TANTRUM_PERIODIC_ENERGY       = 271775,
+    SPELL_CHARGE_SELECTOR               = 260292,
+    SPELL_INDIGESTION                   = 260793,
+    SPELL_TANTRUM_INITIAL               = 260333,
+
+    // Larva
+    SPELL_DESTROY_LARVA                 = 260418,
+    SPELL_METAMORPHOSIS_2               = 260766,
+    SPELL_SUMMON_BLOOD_TICK             = 260353,
+    SPELL_SUMMON_BLOOD_TICK_VISUAL      = 260496,
+
+    // Blood Tick
+    SPELL_BLOOD_BURST_DAMAGE            = 278637,
+    SPELL_SERRATED_FANGS                = 260455,
+
+    // Fetid Maggot
+    SPELL_FEIGN_DEATH                   = 159474
+};
+
+enum CragmawEvents
+{
+    // Cragmaw
+    EVENT_CHARGE_SELECTOR       = 1,
+    EVENT_INDIGESTION,
+    EVENT_NORMAL_REQUEUE,
+
+    // Blood Tick
+    EVENT_SERRATED_FANGS        = 1
+};
+
+enum Points
+{
+    POINT_TANTRUM_START_RND_MOVEMENT    = 0
+};
+
+enum CragmawNPC
+{
+    NPC_FETID_MAGGOT    = 130909
+};
+
+Position const FetidMaggotSpawn = { 857.864f, 984.981f, 39.231f, 4.68147f };
+
+// 131817 - Cragmaw the Infested
+struct boss_cragmaw_the_infested : public BossAI
+{
+    boss_cragmaw_the_infested(Creature* creature) : BossAI(creature, DATA_CRAGMAW_THE_INFESTED) { }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        _JustDied();
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+    }
+
+    void JustAppeared() override
+    {
+        if (instance->GetData(DATA_CRAGMAW_CRAWG_EATING))
+            return;
+
+        DoCastSelf(SPELL_CRAWG_EATING);
+        me->SetEmoteState(EMOTE_STATE_EAT);
+        if (TempSummon* summon = me->SummonCreature(NPC_FETID_MAGGOT, FetidMaggotSpawn))
+        {
+            _fetidMaggotGuid = summon->GetGUID();
+            summon->CastSpell(nullptr, SPELL_FEIGN_DEATH, true);
+            summon->SetUnitFlag(UnitFlags(UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_IMMUNE_TO_NPC | UNIT_FLAG_UNINTERACTIBLE));
```

Comment:
Isn't it best to script SPELL_FEIGN_DEATH to set these flags OnApply and remove them OnRemove?